packages/flutter_tools/lib/src/web/file_generators/main_dart.dart

```diff
@@ -21,16 +22,28 @@ String generateMainDartFile(String appEntrypoint, {
     '',
     "import 'dart:ui_web' as ui_web;",
     "import 'dart:async';",
+    "import 'dart:js_interop';",
     '',
     "import '$appEntrypoint' as entrypoint;",
     "import '$pluginRegistrantEntrypoint' as pluginRegistrant;",
     '',
     'typedef _UnaryFunction = dynamic Function(List<String> args);',
     'typedef _NullaryFunction = dynamic Function();',
     '',
+    '@JS("window")',
+    'external _DomWindow get _domWindow;',
+    '@JS()',
+    '@staticInterop',
+    'class _DomWindow {}',
+    '',
+    'extension _DomWindowExtension on _DomWindow {',
+    "@JS('_flutter_assetManifestBytes')",
+    '  external  set _flutter_assetManifestAsByteList(value);',
+    '}',
     'Future<void> main() async {',
     '  await ui_web.bootstrapEngine(',
     '    runApp: () {',
+    '      _domWindow._flutter_assetManifestAsByteList = [${assetManifestBytes.join(',')}];',
```

Comment:
Can we import `_load_asset_manifest_web.dart` and call a setter on it? Monkey-patching it onto the window feels like a hacky way to establish a dependency between the generated main file and `_load_asset_manifest_web.dart`. The usual way to establish dependencies like this would be to import the library directly.

My second concern is that long-term we'd like the app and engine bootstrap API to be a meaningful clean API surface, so that apps do not need to rely on scary amounts of generated code to bootstrap a Flutter app. Instead, the Flutter app can be "just a Dart library" that you import and run. This means that the bootstrap code should read like something that one would write by hand. This change seems to be taking us in the opposite direction.

One way to clean this up might be to provide a public API for setting the bytes of the manifest. This API would be cross-platform. If the manifest is set manually, the framework will use it, and it will not attempt to load the manifest from a file/bundle/server. If it's not provided, then the framework will fallback to loading from an external asset.

Reply:
I like making the setter public and cross platform, seems like a reasonable approach to take. Though we must document any caching caveats that apply - such that setting the value after the manifest has been read should have no effect.

> My second concern is that long-term we'd like the app and engine bootstrap API to be a meaningful clean API surface

If there is a reasonably detailed proposal for how this might happen, It would be great to share that so we can design around it. Otherwise I don't think it should be the responsibility of this PR to solve the problem of the generated boostraping for web builds.

Reply:
I'm not worried about the generated code, but about the shape of the bootstrap API itself. The complexity of the generated code is merely an indicator of a possible design flaw in the framework API. With this PR as is the framework introduces a hidden API based on monkey-patched `window` object. If that API is not respected, the app will not function. So it is effectively public API (it requires external code to interact with it to initialize the framework), and therefore appropriate to review as part of this PR.

The public setter is my proposal, but maybe @ditman has better ideas. Your concern about repeated invocations of the setter makes sense. Maybe instead of a setter it should be a function named `initializeAssetManifestData` and it could have an assert that ensures it is never called after the asset manifest has been initialized.

Reply:
If you're happy with a public setter I'm happy with a public setter, we just need to document the caveats.

Reply:
Heya, sorry for the silence on this--I was out on vacation at the tail end of last week.

My initial thinking using a hidden JS variable was to avoid adding to Flutter's API surface, but @yjbanov raised a good point in that this change also effectively increases the API surface:

> With this PR as is the framework introduces a hidden API based on monkey-patched window object. If that API is not respected, the app will not function. So it is effectively public API...

While hiding the API reduces the chance of misuse/confusion from most flutter users, it does result in a bunch of complexity that's already been discussed.

I am open to just introducing a public setter.